### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by sauliusgrigaitis** [elliottminns/echo#9]
Trying to compile on Ubuntu 15.10 with `DEVELOPMENT-SNAPSHOT-2016-03-01-a`:

``````
saulius@necolt:~/temp/echo$ swift build
Cloning https://github.com/elliottminns/uv-module.git
Resolved version: 0.1.0
Compiling Swift Module 'Echo' (9 sources)
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
/home/saulius/temp/echo/Packages/Libuv-0.1.0/module.modulemap:3:12: error: header '/usr/local/include/uv.h' not found
    header "/usr/local/include/uv.h"
           ^
/home/saulius/temp/echo/Sources/Server.swift:2:8: error: could not build Objective-C module 'CUV'
import CUV
       ^
<unknown>:0: error: build had 1 command failures
error: exit(1): /home/saulius/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2016-03-01-a/usr/bin/swift-build-tool -f /home/saulius/temp/echo/.build/debug.yaml default```
``````

**Comment by sauliusgrigaitis** [comment]
I installed `libuv` via `apt-get` and fixed header path, but it still doesn't compile:

```
saulius@necolt:~/temp/echo$ swift build
Compiling Swift Module 'Echo' (9 sources)
/home/saulius/temp/echo/Sources/Server.swift:48:45: error: extra argument in call
        uv_ip4_addr("0.0.0.0", Int32(port), bind_addr)
                                            ^~~~~~~~~
/home/saulius/temp/echo/Sources/Server.swift:50:62: error: extra argument in call
        uv_tcp_bind(tcp, UnsafePointer<sockaddr>(bind_addr), 0)
                                                             ^
/home/saulius/temp/echo/Sources/Connection.swift:96:31: error: cannot convert value of type '(UnsafeMutablePointer<uv_handle_t>, size: size_t, buffer: UnsafeMutablePointer<uv_buf_t>) -> ()' to expected argument type 'uv_alloc_cb!'
        uv_read_start(stream, alloc_buffer, on_client_read)
                              ^~~~~~~~~~~~
<unknown>:0: error: build had 1 command failures
error: exit(1): /home/saulius/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2016-03-01-a/usr/bin/swift-build-tool -f /home/saulius/temp/echo/.build/debug.yaml default
```

**Comment by elliottminns** [comment]
Thanks for raising this, the `apt-get` version of libuv is quite outdated and isn't compatible with Blackfish, I recommend installing the one from the github repo by using the below commands:

```
$ sudo apt-get install automake libtool
$ curl -O http://dist.libuv.org/dist/v1.8.0/libuv-v1.8.0.tar.gz
$ tar xzf libuv-v1.8.0.tar.gz
$ cd libuv-v1.8.0 && sh autogen.sh
$ ./configure                                                       
$ make && sudo make install
```

Let me know how that goes.

**Comment by sauliusgrigaitis** [comment]
@elliottminns thanks. I suggest to put this info in README. Just wondering what are benefits using your project comparing to [libdispatch](https://github.com/apple/swift-corelibs-libdispatch)? I would prefer your project if it avoids compilation definitely, but if both needs compilation it's probably better to stick with "more official".

**Comment by elliottminns** [comment]
I will do that now. I've added it in the README for the Blackfish project but must have forgotten to add it for this one. 

The reason I went for libuv over libdispatch is because of the issues that can arise from using multiple threads when dealing with components in a web server layer, such as databases. It's also been battle tested with the use of Node.js as a reliable and scalable platform. 

I agree, the compilation is a pain, however it's a necessary evil **at the moment**. The Swift team are currently adding in the ability to add C based packages as dependencies, which will remove the need to add linker flags. Therefore, I see using libuv as a future investment.
